shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Timelocks.hs

```diff
@@ -97,111 +125,148 @@ decodeVI = RecD ValidityInterval <! From <! From
-pattern TimelockAnd ms <-
-  Timelock (Memo (TimelockAnd' ms) _)
+pattern RequireMOf :: Era era => Int -> StrictSeq (Timelock era) -> Timelock era
+pattern RequireMOf n ms <-
+  Timelock (Memo (MOfN n ms) _)
   where
-    TimelockAnd ms =
-      Timelock $ memoBytes $ Sum TimelockAnd' 2 !> E encodeFoldable ms
+    RequireMOf n ms =
+      Timelock $ memoBytes (encRaw (MOfN n ms))
 
-pattern TimelockOr :: Era era => StrictSeq (Timelock era) -> Timelock era
-pattern TimelockOr ms <-
-  Timelock (Memo (TimelockOr' ms) _)
+pattern RequireTimeExpire :: Era era => StrictMaybe (SlotNo) -> Timelock era
+pattern RequireTimeExpire mslot <-
+  Timelock (Memo (TimeExpire mslot) _)
   where
-    TimelockOr ms =
-      Timelock $ memoBytes (Sum TimelockOr' 3 !> E encodeFoldable ms)
+    RequireTimeExpire mslot =
+      Timelock $ memoBytes (encRaw (TimeExpire mslot))
+
+pattern RequireTimeStart :: Era era => StrictMaybe (SlotNo) -> Timelock era
+pattern RequireTimeStart mslot <-
+  Timelock (Memo (TimeStart mslot) _)
+  where
+    RequireTimeStart mslot =
+      Timelock $ memoBytes (encRaw (TimeStart mslot))
+
+{-# COMPLETE RequireSignature, RequireAllOf, RequireAnyOf, RequireMOf, RequireTimeExpire, RequireTimeStart #-}
 
-{-# COMPLETE Interval, Multi, TimelockAnd, TimelockOr #-}
+-- =================================================================
+-- Evaluating and validating a Timelock
+
+-- PLEASE SOMEONE VERIFY I AM USING atOrAfter and strictlyBefore RIGHT
+
+atOrAfter :: StrictMaybe SlotNo -> StrictMaybe SlotNo -> Bool
+atOrAfter SNothing SNothing = True
+atOrAfter SNothing (SJust _) = True
+atOrAfter (SJust _) SNothing = False
+atOrAfter (SJust i) (SJust j) = i <= j
+
+strictlyBefore :: StrictMaybe SlotNo -> StrictMaybe SlotNo -> Bool
+strictlyBefore SNothing SNothing = True
+strictlyBefore SNothing (SJust _) = False
+strictlyBefore (SJust _) SNothing = True
+strictlyBefore (SJust i) (SJust j) = i < j
+
+evalTimelock ::
+  Era era =>
+  Set (KeyHash 'Witness (Crypto era)) ->
+  ValidityInterval ->
+  Timelock era ->
+  Bool
+evalTimelock _vhks (ValidityInterval mstart _) (RequireTimeStart mslot) =
+  atOrAfter mstart mslot
+evalTimelock _vhks (ValidityInterval _ mexpire) (RequireTimeExpire mslot) =
+  strictlyBefore mslot mexpire
+evalTimelock vhks _vi (RequireSignature hash) = member hash vhks
+evalTimelock vhks vi (RequireAllOf xs) =
+  all (evalTimelock vhks vi) xs
+evalTimelock vhks vi (RequireAnyOf xs) =
+  any (evalTimelock vhks vi) xs
+evalTimelock vhks vi (RequireMOf m xs) =
+  m <= sum (fmap (\x -> if evalTimelock vhks vi x then 1 else 0) xs)
 
 -- =========================================================
 -- Operations on Timelock scripts
 
 inInterval :: SlotNo -> ValidityInterval -> Bool
 inInterval _slot (ValidityInterval SNothing SNothing) = True
-inInterval slot (ValidityInterval SNothing (SJust i_f)) = slot <= i_f
+inInterval slot (ValidityInterval SNothing (SJust i_f)) = slot < i_f -- An Open interval uses (<) rather then (<=)
```

Comment:
Rather than having these comments here, they would be more useful on the haddock for `ValidityInterval` itself (or perhaps also on the haddock for this function). That way, it becomes more clear to people who are _using_ this function what semantics it is embodying.

Reply:
Done